[AMDGPU] Use `bf16` instead of `i16` for bfloat

shiltian opened the pull request:
Currently we generally use `i16` to represent `bf16` in those tablegen files. This patch is trying to use `bf16` directly.

Fix #79369.

rampitec commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -2819,11 +2819,11 @@ def int_amdgcn_fdot2_f16_f16 :
   ClangBuiltin<"__builtin_amdgcn_fdot2_bf16_bf16">,
   DefaultAttrsIntrinsic<
-    [llvm_i16_ty],   // %r
+    [llvm_bfloat_ty],   // %r
```
This is the change in the existing interface, it will break programs already using it. It also will break clang builtin.

You can get away with changing only instruction's profile to use BF16 and tweak pattern with casts to bf16.

arsenm replied:
It will just maybe require changing the clang builtin codegen to insert the casts. You still need to make this change

arsenm replied:
You should also add IR auto upgrade

shiltian replied:
Does it make sense if we also update both the Clang builtins and LLVM builtins here, and use IR auto upgrade to rewrite existing uses?

arsenm replied:
Changing the clang builtin breaks user code. Anyone using it will now have an implicit itofp cast to bfloat. For the first step, I would leave the clang builtin alone (and only make the codegen change to emit the new IR). We would either need to introduce a new builtin with the corrected signature, or add a language flag to change the behavior

rampitec replied:
clang/test/CodeGenOpenCL/builtins-amdgcn-dl-insts-gfx11.cl fails. You need to insert casts to bf16 while lowering it to make it working.

shiltian replied:
The cast will be inserted automatically in `clang/lib/CodeGen/CGBuiltin.cpp` after removing the two assertions. I reverted my change to the test case by accident. Lol

shiltian commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -1562,8 +1562,9 @@ bool IRTranslator::translateBitCast(const User &U,
                                  MachineIRBuilder &MIRBuilder) {
-  if (U.getType()->getScalarType()->isBFloatTy() ||
-      U.getOperand(0)->getType()->getScalarType()->isBFloatTy())
+  if (Opcode != TargetOpcode::G_BITCAST &&
```
I think bitcast can still be supported even if we don't know how to convert from bfloat to other types mentioned in #71470. If so, we might want a separate patch for this.

arsenm replied:
Unrelated, but this also misses hidden casts implied by the ABI

arsenm replied:
This change should be done separately

rampitec replied:
This is actually an orthogonal problem. Global ISel is completely broken for bf16 and whatever the outcome of the supporting bf16 in codegen is we just need to be ready some gisel tests will fail and will need to be disabled.

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -2835,8 +2835,8 @@ def int_amdgcn_fdot2_f32_bf16 :
     [
-      llvm_v2i16_ty, // %a
-      llvm_v2i16_ty, // %b
+      llvm_v2bf16_ty, // %a
+      llvm_v2bf16_ty, // %b
```
For potential revert pain minimization, I would do each intrinsic in a separate commit

arsenm commented on llvm/test/MC/AMDGPU/bf16_imm.s:
```diff
@@ -0,0 +1,8 @@
+// RUN: llvm-mc -arch=amdgcn -mcpu=gfx1100 -show-encoding %s | FileCheck %s
+// RUN: llvm-mc -arch=amdgcn -mcpu=gfx1200 -show-encoding %s | FileCheck %s
+
+v_dot2_bf16_bf16 v5, v1, v2, 100.0
```
does this help with #79369 at all?

shiltian commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -5908,8 +5908,6 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
           }
         }
 
-        assert(ArgValue->getType()->canLosslesslyBitCastTo(PTy) &&
```
This change might need to go to a separate patch.

rampitec commented on llvm/test/MC/AMDGPU/bf16_imm.s:
```diff
@@ -0,0 +1,8 @@
+// CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0xc8,0x42,0x00,0x00]
+
+v_dot2_bf16_bf16 v5, v1, v2, 1.0
+// v_dot2_bf16_bf16 v5, v1, v2, 0x3f80 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0x80,0x3f,0x00,0x00]
```
FYI: this shall be inline literal. I.e:
    0xd6672005
    0x03ca0501

rampitec commented on llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp:
```diff
@@ -521,8 +521,11 @@ void AMDGPUInstPrinter::printImmediateV216(uint32_t Imm, uint8_t OpType,
   case AMDGPU::OPERAND_REG_IMM_V2FP16:
+  case AMDGPU::OPERAND_REG_INLINE_C_V2BF16:
   case AMDGPU::OPERAND_REG_INLINE_C_V2FP16:
+  case AMDGPU::OPERAND_REG_INLINE_AC_V2BF16:
```
It does not seem right, and there are no tests for v2bf16 added. I am not sure though we have instructions which can accept this type of operand.

rampitec commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4181,13 +4181,20 @@ bool SIInstrInfo::isInlineConstant(const MachineOperand &MO,
   case AMDGPU::OPERAND_REG_INLINE_C_V2INT16:
   case AMDGPU::OPERAND_REG_INLINE_AC_V2INT16:
     return AMDGPU::isInlinableLiteralV2I16(Imm);
+  case AMDGPU::OPERAND_REG_IMM_V2BF16:
```
It does not seem isInlinableLiteralV2F16() can handle bf16.

shiltian replied:
Yeah, I made some mistakes here. Will take care of them.

rampitec commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:
```diff
@@ -79,17 +79,17 @@ define amdgpu_ps void @test_llvm_amdgcn_fdot2_bf16_bf16_sis(
 ; GFX11-NEXT:    v_mov_b32_e32 v2, s1
 ; GFX11-NEXT:    s_delay_alu instid0(VALU_DEP_1)
-; GFX11-NEXT:    v_dot2_bf16_bf16 v2, s0, 0x10001, v2
+; GFX11-NEXT:    v_dot2_bf16_bf16 v2, s0, 0x3f803f80, v2
```
This shall be encoded as inline immediate 1.0.

shiltian replied:
Yeah, but I tested the FP16 version `llvm.amdgcn.fdot2.f16.f16` (w/ trunk w/o my patch), it generates `v_dot2_f16_f16 v2, s0, 0x3c003c00, v2`. I think we generally have issues with showing inline literals.

rampitec replied:
Well, this is unrelated to the patch itself. We can use inline 1.0 here, but then we must use op_sel_hi to produce it in the high half.

shiltian replied:
FWIW, #81345 can solve the issue but I'm struggling with getting two test cases passed.

shiltian commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:
```diff
@@ -2730,6 +2749,12 @@ std::optional<unsigned> getInlineEncodingV2I16(uint32_t Literal) {
+// Encoding of the literal as an inline constant for a V_PK_*_BF16 instruction
+// or nullopt.
+std::optional<unsigned> getInlineEncodingV2BF16(uint32_t Literal) {
+  return getInlineEncodingV216(true, Literal);
```
This part is still WIP along with https://github.com/llvm/llvm-project/pull/81345.

shiltian commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:
```diff
@@ -2660,15 +2660,34 @@ bool isInlinableLiteral16(int16_t Literal, bool HasInv2Pi) {
-         Val == 0x3118;   // 1/2pi
+
+  // FP16
+  if (Val == 0x3C00 || // 1.0
```
This function might be removed eventually in https://github.com/llvm/llvm-project/pull/81345.

shiltian commented:
The patch is in a good shape now. I have made two other prime patches (#81674 and #81669). I'll rebase this one once they are landed.

This patch only changes one bf16 instruction with the necessary infrastructure for others. I'll update all of them once this patch is landed.

However, I don't think `isInlinableLiteral16` works correctly because the encoding of the floating point inline literals are different for fp16 and bf16, but apparently for now it can only recognize fp16 encoding. This patch at least makes the asm printer work properly. #81345 is trying to fix it correctly, but that is unrelated to this patch.

shiltian commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll:
```diff
@@ -1,8 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
 ; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=GFX11,SDAG-GFX11
-; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx1100 -verify-machineinstrs < %s | FileCheck %s --check-prefixes=GFX11,GISEL-GFX11
```
This change is because of the discussion here (https://github.com/llvm/llvm-project/pull/80908/files#r1483394728).

rampitec replied:
Change 'RUN' with 'XUN' and add a comment instead.

rampitec commented on llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUInstPrinter.cpp:
```diff
@@ -488,6 +488,49 @@ static bool printImmediateFloat16(uint32_t Imm, const MCSubtargetInfo &STI,
+  if (printImmediateBFloat16(HImm, STI, O))
+    return;
+
+  uint64_t Imm16 = static_cast<uint16_t>(Imm);
```
It's the same as HImm above.

shiltian replied:
Yeah, but it is to make the type promoted to `uint64_t` w/o any ambiguity.

rampitec commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4185,9 +4185,17 @@ bool SIInstrInfo::isInlineConstant(const MachineOperand &MO,
   case AMDGPU::OPERAND_REG_IMM_FP16_DEFERRED:
+  case AMDGPU::OPERAND_REG_INLINE_C_BF16:
   case AMDGPU::OPERAND_REG_INLINE_C_FP16:
+  case AMDGPU::OPERAND_REG_INLINE_AC_BF16:
```
It seems isInlinableLiteral16() cannot handle bf16?

shiltian replied:
Yes, `isInlinableLiteral16` can't because it can't tell `fp16` and `bf16` apart by just looking at the value. That's the reason I tried really hard to get rid of `isInlinableLiteral16` in #81345 and favors the explicit version.

rampitec replied:
But right in this place you know the actual format. So you can split F16 and BF16 code and call different functions.

shiltian replied:
Okay, I copied `isInlinableLiteralBF16` from https://github.com/llvm/llvm-project/pull/81345. Initially I thought https://github.com/llvm/llvm-project/pull/81345 would be landed before this one, but I went down that rabbit hole and still didn't make it fully work.

rampitec commented on llvm/test/MC/AMDGPU/bf16_imm.s:
```diff
@@ -0,0 +1,8 @@
+// RUN: llvm-mc -arch=amdgcn -mcpu=gfx1100 -show-encoding %s | FileCheck %s
```
You also need a disasm test for this.

shiltian commented on llvm/test/MC/Disassembler/AMDGPU/bf16_imm.txt:
```diff
@@ -0,0 +1,8 @@
+# CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8
+0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0xc8,0x42,0x00,0x00
+
+# CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x3c00
```
The disassembler doesn't work properly because when it sees `242` with bitwidth 16, it doesn't know whether it is bf16 or fp16.

rampitec commented on llvm/test/MC/Disassembler/AMDGPU/bf16_imm.txt:
```diff
@@ -0,0 +1,8 @@
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1100 -disassemble -show-encoding < %s | FileCheck %s
+# RUN: llvm-mc -triple=amdgcn -mcpu=gfx1200 -disassemble -show-encoding < %s | FileCheck %s
+
+# CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8
```
Add encoding to the check lines. Currently it is broken and encoded value is different from decoded one.

shiltian commented:
I'll create a ticket about the decoder.

jayfoad commented on llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCCodeEmitter.cpp:
```diff
@@ -157,6 +157,27 @@ static uint32_t getLit16Encoding(uint16_t Val, const MCSubtargetInfo &STI) {
+    return IntImm;
+
+  // clang-format off
+  switch (Val) {
```
Can this call `getInlineEncodingV2BF16`?

shiltian replied:
In theory, yes, but for now we can't because `getInlineEncodingV2BF16` can't handle some cases (that I didn't dig yet). It looks like in the conversion between `uint16_t` and `uint32_t` that makes some test cases fail. IMO we need to unify them (not only for 16-bit) in one place instead of having almost the same logic at least in three places.

jayfoad replied:
Yeah, I really don't like having 4 different copies of this list of hex values (0x3f00, 0xbf00...).

rampitec commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:
```diff
@@ -2652,6 +2652,23 @@ bool isInlinableLiteral32(int32_t Literal, bool HasInv2Pi) {
 
+bool isInlinableLiteralBF16(int16_t Literal, bool HasInv2Pi) {
+  if (!HasInv2Pi)
+    return false;
```
It does not change the behavior, but generally it shall only matter when you compare value to 0x3E22.

rampitec commented on llvm/test/MC/AMDGPU/bf16_imm.s:
```diff
@@ -0,0 +1,8 @@
+// CHECK: v_dot2_bf16_bf16 v5, v1, v2, 0x42c8 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xfe,0x03,0xc8,0x42,0x00,0x00]
+
+v_dot2_bf16_bf16 v5, v1, v2, 1.0
+// CHECK: v_dot2_bf16_bf16 v5, v1, v2, 1.0 ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0xca,0x03]
```
Can you add couple more tests here? The same instruction, but with the immediate in place of v2bf16 operand, so we test that code path too. Both immediate and inline literal. Here and in disasm.

shiltian replied:
It looks like our assembler doesn't like the case such as `v_dot2_bf16_bf16 v5, v1, 0x42c842c8, 0x42c8`, even w/o this patch. Even the asm (`v_dot2_bf16_bf16 v2, s0, 0x3f803f80, v2`) generated in another file above (`llvm/test/CodeGen/AMDGPU/llvm.amdgcn.fdot2.bf16.bf16.ll`) can't be recognized by our assembler. 😄

shiltian committed:
[RFC][WIP][AMDGPU] Use `bf16` instead of `i16` for bfloat

Currently it looks like we generally use `i16` to represent `bf16` in those tablegen
files. I'm not sure of the reason behind it. My wild guess is the type `bf16` was
not available when we enabled the support. This patch is trying to use `bf16`
directly in those tablegen files, aiming at fixing #79369. Of course for #79369
a workaround can be to treat all `INT16` variants as `BFloat` in `getOpFltSemantics`,
but it doesn't look good IMHO.

Since I'm fairly new to AMDGPU backend, I'd appreciate it if you can point out
where I don't understand correctly.

rampitec reviewed (approved):
Thanks. There are definitely at least 2 outstanding problems, but it seems there are no regressions comparing to what we have now. LGTM.
